spark/src/main/scala/org/apache/spark/ml/regression/SmileRegressor.scala

```diff
@@ -0,0 +1,241 @@
+/*******************************************************************************
+ * Copyright (c) 2010-2020 Haifeng Li. All rights reserved.
+ *
+ * Smile is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version.
+ *
+ * Smile is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Smile.  If not, see <https://www.gnu.org/licenses/>.
+ ******************************************************************************/
+
+package org.apache.spark.ml.regression
+
+import java.io.{ObjectInputStream, ObjectOutputStream}
+
+import org.apache.hadoop.fs.{FileSystem, Path}
+import org.apache.spark.SparkContext
+import org.apache.spark.ml.linalg.Vector
+import org.apache.spark.ml.param._
+import org.apache.spark.ml.util.Instrumentation._
+import org.apache.spark.ml.util.{Identifiable, _}
+import org.apache.spark.ml.{Predictor, PredictorParams}
+import org.apache.spark.sql.Dataset
+import org.apache.spark.storage.StorageLevel
+import org.json4s.JObject
+import org.json4s.JsonDSL._
+import org.json4s.jackson.JsonMethods._
+
+private[ml] trait SmileRegressorParams extends Params with PredictorParams {
+
+  type Trainer =
+    (Array[Array[Double]], Array[Double]) => smile.regression.Regression[Array[Double]]
+
+  /**
+   * smile regressor used for training
+   *
+   * @group param
+   */
+  val trainer: Param[Trainer] =
+    new Param[Trainer](this, "trainer", "smile regressor used for training")
+
+  /** @group getParam */
+  def getTrainer: Trainer = $(trainer)
+
+}
+
+private[ml] object SmileRegressorParams {
+
+  def saveImpl(
+                instance: SmileRegressorParams,
+                path: String,
+                sc: SparkContext,
+                extraMetadata: Option[JObject] = None): Unit = {
+
+    val params = instance.extractParamMap().toSeq
+    val jsonParams = render(
+      params
+        .filter { case ParamPair(p, _) => p.name != "trainer" }
+        .map { case ParamPair(p, v) => p.name -> parse(p.jsonEncode(v)) }
+        .toList)
+
+    DefaultParamsWriter.saveMetadata(instance, path, sc, extraMetadata, Some(jsonParams))
+  }
+
+  def loadImpl(
+                path: String,
+                sc: SparkContext,
+                expectedClassName: String): DefaultParamsReader.Metadata = {
+
+    DefaultParamsReader.loadMetadata(path, sc, expectedClassName)
+  }
+
+}
+
+class SmileRegressor(override val uid: String)
```

Comment:
Regressor usually refers to the independent variables (X), not the model. It is better named as SmileRegression.

Reply:
That's the usual naming convention in Spark MLLib with RandomForestRegressor and FMRegressor : https://spark.apache.org/docs/latest/ml-classification-regression.html#factorization-machines-regressor but as SurvivalRegression or IsotonicRegression is used to i'm going to change it as SmileRegression.